File: context/portfolio_ledger.py
from __future__ import annotations

from collections import defaultdict
from copy import deepcopy
from datetime import datetime, timezone

from context.portfolio_fee_registry import estimate_transaction_charges


DEFAULT_CURRENCY = "INR"


def _as_string(value, default: str = "") -> str:
    if value is None:
        return default
    return str(value).strip()


def _as_float(value, default: float = 0.0) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return default


def _normalize_charge_snapshot(value) -> dict:
    row = value if isinstance(value, dict) else {}
    totals = row.get("totals") if isinstance(row.get("totals"), dict) else {}
    return {
        "platformId": _as_string(row.get("platformId")),
        "platformLabel": _as_string(row.get("platformLabel")),
        "segmentKey": _as_string(row.get("segmentKey")),
        "segmentLabel": _as_string(row.get("segmentLabel")),
        "side": _as_string(row.get("side")).upper() or "BUY",
        "turnover": _as_float(row.get("turnover"), 0.0),
        "lines": [
            {
                "key": _as_string(line.get("key")),
                "label": _as_string(line.get("label")),
                "amount": _as_float(line.get("amount"), 0.0),
            }
            for line in (row.get("lines") or [])
            if isinstance(line, dict)
        ],
        "totals": {str(key): _as_float(val, 0.0) for key, val in totals.items()},
        "totalCharges": _as_float(row.get("totalCharges"), 0.0),
        "sourceTitle": _as_string(row.get("sourceTitle")),
        "sourceUrl": _as_string(row.get("sourceUrl")),
        "exactness": _as_string(row.get("exactness")),
        "registryVersion": _as_string(row.get("registryVersion")),
    }


def normalize_portfolio_transaction(row: dict) -> dict | None:
    if not isinstance(row, dict):
        return None
    symbol = _as_string(row.get("symbol")).upper()
    if not symbol:
        return None
    side = _as_string(row.get("side") or row.get("transactionType") or "BUY").upper()
    if side not in {"BUY", "SELL", "DIVIDEND", "FEE", "TAX", "ADJUSTMENT"}:
        side = "BUY"
    price = _as_float(row.get("price"), _as_float(row.get("buyPrice"), 0.0))
    quantity = _as_float(row.get("quantity"), 0.0)
    trade_date = _as_string(row.get("tradeDate") or row.get("purchaseDate"))
    charge_snapshot = _normalize_charge_snapshot(row.get("chargeSnapshot"))
    if not charge_snapshot.get("lines") and side in {"BUY", "SELL"} and quantity > 0 and price > 0:
        charge_snapshot = estimate_transaction_charges(
            {
                "platform": row.get("platform"),
                "segment": row.get("segment"),
                "purchaseType": row.get("purchaseType"),
                "state": row.get("state"),
                "side": side,
                "quantity": quantity,
                "price": price,
                "manualCharge": row.get("manualCharge"),
                "manualTax": row.get("manualTax"),
            }
        )
    current_price = _as_float(row.get("currentPrice"), price)
    return {
        "id": _as_string(row.get("id")) or f"{symbol}_{int(datetime.now(timezone.utc).timestamp() * 1000)}",
        "entryType": "transaction",
        "side": side,
        "transactionSubtype": _as_string(row.get("transactionSubtype")),
        "symbol": symbol,
        "assetName": _as_string(row.get("assetName")) or symbol,
        "description": _as_string(row.get("description")),
        "notes": _as_string(row.get("notes")),
        "brokerReference": _as_string(row.get("brokerReference")),
        "importSource": _as_string(row.get("importSource")),
        "importBatchId": _as_string(row.get("importBatchId")),
        "purchaseType": _as_string(row.get("purchaseType")) or "Delivery",
        "segment": _as_string(row.get("segment")) or "Equity",
        "tradeDate": trade_date,
        "platform": _as_string(row.get("platform")),
        "country": _as_string(row.get("country")) or "India",
        "state": _as_string(row.get("state")),
        "quantity": quantity,
        "price": price,
        "currentPrice": current_price,
        "currencySymbol": _as_string(row.get("currencySymbol")) or DEFAULT_CURRENCY,
        "manualCharge": _as_float(row.get("manualCharge"), 0.0),
        "manualTax": _as_float(row.get("manualTax"), 0.0),
        "chargeSnapshot": charge_snapshot,
        "createdAt": _as_string(row.get("createdAt")) or datetime.now(timezone.utc).isoformat(),
        "legacyImported": bool(row.get("legacyImported")),
    }


def migrate_legacy_position(row: dict) -> dict | None:
    if not isinstance(row, dict):
        return None
    symbol = _as_string(row.get("symbol")).upper()
    if not symbol:
        return None
    quantity = _as_float(row.get("quantity"), 0.0)
    price = _as_float(row.get("buyPrice"), 0.0)
    if quantity <= 0 or price <= 0:
        return None
    return normalize_portfolio_transaction(
        {
            "id": _as_string(row.get("id")) or f"legacy_{symbol}",
            "symbol": symbol,
            "assetName": _as_string(row.get("assetName")) or symbol,
            "description": row.get("description"),
            "notes": row.get("notes"),
            "purchaseType": row.get("purchaseType") or "Delivery",
            "segment": row.get("segment") or "Equity",
            "tradeDate": row.get("purchaseDate"),
            "platform": row.get("platform"),
            "country": row.get("country") or "India",
            "state": row.get("state"),
            "quantity": quantity,
            "price": price,
            "currentPrice": _as_float(row.get("currentPrice"), price),
            "currencySymbol": row.get("currencySymbol") or DEFAULT_CURRENCY,
            "manualCharge": 0.0,
            "manualTax": 0.0,
            "chargeSnapshot": {
                "platformId": "",
                "platformLabel": _as_string(row.get("platform")),
                "segmentKey": "",
                "segmentLabel": "",
                "side": "BUY",
                "turnover": round(quantity * price, 2),
                "lines": [],
                "totals": {},
                "totalCharges": 0.0,
                "sourceTitle": "Legacy migrated holding",
                "sourceUrl": "",
                "exactness": "legacy_import",
                "registryVersion": "",
            },
            "legacyImported": True,
        }
    )


def clean_portfolio_entries(rows: list | None) -> list[dict]:
    cleaned: list[dict] = []
    seen_ids: set[str] = set()
    for row in rows or []:
        candidate = normalize_portfolio_transaction(row) if isinstance(row, dict) and str(row.get("entryType") or row.get("side") or row.get("transactionType")).strip() else migrate_legacy_position(row)
        if not candidate:
            continue
        row_id = candidate["id"]
        if row_id in seen_ids:
            candidate["id"] = f"{row_id}_{len(cleaned) + 1}"
        seen_ids.add(candidate["id"])
        cleaned.append(candidate)
    return cleaned


def clean_portfolios(portfolios: dict | None) -> dict[str, list[dict]]:
    out: dict[str, list[dict]] = {}
    if isinstance(portfolios, dict):
        for name, rows in portfolios.items():
            portfolio_name = _as_string(name)
            if not portfolio_name:
                continue
            out[portfolio_name] = clean_portfolio_entries(rows if isinstance(rows, list) else [])
    if not out:
        out["Main"] = []
    return out


def _charge_total(txn: dict) -> float:
    return _as_float(((txn.get("chargeSnapshot") or {}).get("totalCharges")), 0.0)


def derive_holdings_from_transactions(transactions: list[dict] | None) -> list[dict]:
    by_symbol: dict[str, dict] = {}
    ordered = sorted(
        [txn for txn in (transactions or []) if isinstance(txn, dict)],
        key=lambda row: (_as_string(row.get("tradeDate") or row.get("createdAt")), _as_string(row.get("createdAt")), _as_string(row.get("id"))),
    )
    for txn in ordered:
        side = _as_string(txn.get("side")).upper()
        symbol = _as_string(txn.get("symbol")).upper()
        if not symbol:
            continue
        bucket = by_symbol.setdefault(
            symbol,
            {
                "symbol": symbol,
                "assetName": _as_string(txn.get("assetName")) or symbol,
                "description": _as_string(txn.get("description")),
                "notes": _as_string(txn.get("notes")),
                "purchaseType": _as_string(txn.get("purchaseType")) or "Delivery",
                "segment": _as_string(txn.get("segment")) or "Equity",
                "platform": _as_string(txn.get("platform")),
                "country": _as_string(txn.get("country")) or "India",
                "currencySymbol": _as_string(txn.get("currencySymbol")) or DEFAULT_CURRENCY,
                "currentPrice": _as_float(txn.get("currentPrice"), _as_float(txn.get("price"), 0.0)),
                "quantity": 0.0,
                "costBasis": 0.0,
                "realizedPnl": 0.0,
                "buyCharges": 0.0,
                "sellCharges": 0.0,
                "totalChargesPaid": 0.0,
                "transactionCount": 0,
                "lastTradeDate": "",
            },
        )
        bucket["assetName"] = _as_string(txn.get("assetName")) or bucket["assetName"]
        bucket["description"] = _as_string(txn.get("description")) or bucket["description"]
        bucket["notes"] = _as_string(txn.get("notes")) or bucket["notes"]
        bucket["purchaseType"] = _as_string(txn.get("purchaseType")) or bucket["purchaseType"]
        bucket["segment"] = _as_string(txn.get("segment")) or bucket["segment"]
        bucket["platform"] = _as_string(txn.get("platform")) or bucket["platform"]
        bucket["country"] = _as_string(txn.get("country")) or bucket["country"]
        bucket["currencySymbol"] = _as_string(txn.get("currencySymbol")) or bucket["currencySymbol"]
        bucket["currentPrice"] = _as_float(txn.get("currentPrice"), bucket["currentPrice"])
        bucket["lastTradeDate"] = _as_string(txn.get("tradeDate") or txn.get("createdAt")) or bucket["lastTradeDate"]
        bucket["transactionCount"] += 1
        quantity = _as_float(txn.get("quantity"), 0.0)
        price = _as_float(txn.get("price"), 0.0)
        charges = _charge_total(txn)
        bucket["totalChargesPaid"] += charges
        if side == "BUY":
            bucket["quantity"] += quantity
            bucket["costBasis"] += (quantity * price) + charges
            bucket["buyCharges"] += charges
        elif side == "SELL":
            sell_qty = min(quantity, bucket["quantity"])
            avg_cost = bucket["costBasis"] / bucket["quantity"] if bucket["quantity"] > 0 else 0.0
            cost_removed = avg_cost * sell_qty
            net_proceeds = (sell_qty * price) - charges
            bucket["realizedPnl"] += net_proceeds - cost_removed
            bucket["quantity"] = max(bucket["quantity"] - sell_qty, 0.0)
            bucket["costBasis"] = max(bucket["costBasis"] - cost_removed, 0.0)
            bucket["sellCharges"] += charges
        elif side == "DIVIDEND":
            bucket["realizedPnl"] += (quantity * price) - charges
        elif side in {"FEE", "TAX"}:
            bucket["realizedPnl"] -= charges if charges > 0 else quantity * price
        elif side == "ADJUSTMENT":
            subtype = _as_string(txn.get("transactionSubtype")).lower()
            if subtype == "split" and quantity > 0:
                bucket["quantity"] += quantity
            elif subtype == "bonus" and quantity > 0:
                bucket["quantity"] += quantity
            else:
                bucket["costBasis"] += quantity * price

    holdings: list[dict] = []
    for symbol, bucket in by_symbol.items():
        quantity = bucket["quantity"]
        avg_cost = bucket["costBasis"] / quantity if quantity > 0 else 0.0
        market_value = quantity * bucket["currentPrice"]
        exit_preview = estimate_transaction_charges(
            {
                "platform": bucket["platform"],
                "segment": bucket["segment"],
                "purchaseType": bucket["purchaseType"],
                "side": "SELL",
                "quantity": quantity,
                "price": bucket["currentPrice"],
            }
        ) if quantity > 0 and bucket["currentPrice"] > 0 else {"totalCharges": 0.0, "totals": {}, "lines": []}
        projected_exit = _as_float(exit_preview.get("totalCharges"), 0.0)
        gross_pnl = market_value - bucket["costBasis"]
        net_pnl = gross_pnl - projected_exit
        holdings.append(
            {
                "id": symbol,
                "symbol": symbol,
                "assetName": bucket["assetName"],
                "description": bucket["description"],
                "notes": bucket["notes"],
                "purchaseType": bucket["purchaseType"],
                "segment": bucket["segment"],
                "platform": bucket["platform"],
                "country": bucket["country"],
                "currencySymbol": bucket["currencySymbol"],
                "quantity": round(quantity, 6),
                "buyPrice": round(avg_cost, 4),
                "averageCost": round(avg_cost, 4),
                "invested": round(bucket["costBasis"], 2),
                "currentPrice": round(bucket["currentPrice"], 4),
                "current": round(market_value, 2),
                "grossPnl": round(gross_pnl, 2),
                "netPnl": round(net_pnl, 2),
                "projectedExitCharges": round(projected_exit, 2),
                "realizedPnl": round(bucket["realizedPnl"], 2),
                "totalChargesPaid": round(bucket["totalChargesPaid"], 2),
                "transactionCount": bucket["transactionCount"],
                "lastTradeDate": bucket["lastTradeDate"],
                "exitChargePreview": deepcopy(exit_preview),
            }
        )
    holdings.sort(key=lambda row: (-row["current"], row["symbol"]))
    return holdings


def derive_portfolio_stats(transactions: list[dict] | None) -> dict:
    rows = derive_holdings_from_transactions(transactions)
    by_segment: dict[str, dict] = defaultdict(lambda: {"invested": 0.0, "current": 0.0, "net": 0.0})
    by_platform: dict[str, dict] = defaultdict(lambda: {"invested": 0.0, "current": 0.0, "net": 0.0})
    by_country: dict[str, dict] = defaultdict(lambda: {"invested": 0.0, "current": 0.0, "net": 0.0})
    invested = current = realized = gross = net = projected_exit = charges_paid = 0.0
    profitable = 0
    for row in rows:
        invested += row["invested"]
        current += row["current"]
        realized += row["realizedPnl"]
        gross += row["grossPnl"]
        net += row["netPnl"]
        projected_exit += row["projectedExitCharges"]
        charges_paid += row["totalChargesPaid"]
        profitable += 1 if row["netPnl"] >= 0 else 0
        segment = row["segment"] or "Other"
        platform = row["platform"] or "Unspecified"
        country = row["country"] or "Unspecified"
        by_segment[segment]["invested"] += row["invested"]
        by_segment[segment]["current"] += row["current"]
        by_segment[segment]["net"] += row["netPnl"]
        by_platform[platform]["invested"] += row["invested"]
        by_platform[platform]["current"] += row["current"]
        by_platform[platform]["net"] += row["netPnl"]
        by_country[country]["invested"] += row["invested"]
        by_country[country]["current"] += row["current"]
        by_country[country]["net"] += row["netPnl"]
    recent_transactions = sorted(
        [normalize_portfolio_transaction(txn) for txn in transactions or [] if normalize_portfolio_transaction(txn)],
        key=lambda row: (_as_string(row.get("tradeDate") or row.get("createdAt")), _as_string(row.get("createdAt"))),
        reverse=True,
    )[:10]

    def _map_rows(grouped: dict[str, dict]) -> list[dict]:
        return [
            {
                "label": label,
                "invested": round(values["invested"], 2),
                "current": round(values["current"], 2),
                "pnl": round(values["current"] - values["invested"], 2),
                "net": round(values["net"], 2),
            }
            for label, values in grouped.items()
        ]

    return {
        "holdings": len(rows),
        "profitable": profitable,
        "invested": round(invested, 2),
        "current": round(current, 2),
        "realizedPnl": round(realized, 2),
        "grossPnl": round(gross, 2),
        "netAfterCosts": round(net, 2),
        "projectedExitCharges": round(projected_exit, 2),
        "totalChargesPaid": round(charges_paid, 2),
        "bySegment": _map_rows(by_segment),
        "byPlatform": _map_rows(by_platform),
        "byCountry": _map_rows(by_country),
        "recentPurchases": recent_transactions,
        "holdingsRows": rows,
    }

File: stock-analysis-dashboard/src/portfolioLedger.js
const round = (value, digits = 2) => Number(Number(value || 0).toFixed(digits));

const asString = (value, fallback = '') => (value == null ? fallback : String(value).trim());
const asNumber = (value, fallback = 0) => {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
};

export const resolveSegmentKey = (purchaseType = '', segment = '') => {
  const purchase = asString(purchaseType);
  if (purchase === 'Delivery') return 'equity_delivery';
  if (purchase === 'Intraday') return 'equity_intraday';
  if (purchase === 'Futures') return 'futures';
  if (purchase === 'Options') return 'options';
  if (purchase === 'ETF') return 'etf_delivery';
  if (purchase === 'Mutual Fund') return 'mutual_fund';
  const normalizedSegment = asString(segment).toLowerCase();
  if (normalizedSegment === 'etf') return 'etf_delivery';
  if (normalizedSegment === 'mutual fund') return 'mutual_fund';
  if (normalizedSegment === 'commodity') return 'commodity_futures';
  if (normalizedSegment === 'fx') return 'currency';
  return 'equity_delivery';
};

const normalizeChargeSnapshot = (value) => {
  const row = value && typeof value === 'object' ? value : {};
  const totals = row.totals && typeof row.totals === 'object' ? row.totals : {};
  return {
    platformId: asString(row.platformId),
    platformLabel: asString(row.platformLabel),
    segmentKey: asString(row.segmentKey),
    segmentLabel: asString(row.segmentLabel),
    side: asString(row.side || 'BUY').toUpperCase(),
    turnover: asNumber(row.turnover, 0),
    lines: Array.isArray(row.lines)
      ? row.lines.map((line) => ({
        key: asString(line?.key),
        label: asString(line?.label),
        amount: asNumber(line?.amount, 0),
      }))
      : [],
    totals: Object.fromEntries(Object.entries(totals).map(([key, val]) => [key, asNumber(val, 0)])),
    totalCharges: asNumber(row.totalCharges, 0),
    sourceTitle: asString(row.sourceTitle),
    sourceUrl: asString(row.sourceUrl),
    exactness: asString(row.exactness),
    registryVersion: asString(row.registryVersion),
  };
};

export const normalizePortfolioTransaction = (row) => {
  if (!row || typeof row !== 'object') return null;
  const symbol = asString(row.symbol).toUpperCase();
  if (!symbol) return null;
  const price = asNumber(row.price ?? row.buyPrice, 0);
  const quantity = asNumber(row.quantity, 0);
  const side = asString(row.side || row.transactionType || 'BUY').toUpperCase() || 'BUY';
  return {
    id: asString(row.id || `${Date.now()}_${symbol}`),
    entryType: 'transaction',
    side,
    transactionSubtype: asString(row.transactionSubtype),
    symbol,
    assetName: asString(row.assetName || row.name || symbol) || symbol,
    description: asString(row.description),
    notes: asString(row.notes),
    brokerReference: asString(row.brokerReference),
    importSource: asString(row.importSource),
    importBatchId: asString(row.importBatchId),
    purchaseType: asString(row.purchaseType || 'Delivery') || 'Delivery',
    tradeDate: asString(row.tradeDate || row.purchaseDate),
    platform: asString(row.platform),
    country: asString(row.country || 'India') || 'India',
    state: asString(row.state),
    segment: asString(row.segment || 'Equity') || 'Equity',
    quantity,
    price,
    currentPrice: asNumber(row.currentPrice, price),
    currencySymbol: asString(row.currencySymbol || 'INR') || 'INR',
    manualCharge: asNumber(row.manualCharge, 0),
    manualTax: asNumber(row.manualTax, 0),
    chargeSnapshot: normalizeChargeSnapshot(row.chargeSnapshot),
    createdAt: asString(row.createdAt || new Date().toISOString()),
    legacyImported: Boolean(row.legacyImported),
  };
};

export const migrateLegacyPosition = (row) => {
  if (!row || typeof row !== 'object') return null;
  const symbol = asString(row.symbol).toUpperCase();
  if (!symbol) return null;
  const quantity = asNumber(row.quantity, 0);
  const price = asNumber(row.buyPrice, 0);
  if (quantity <= 0 || price <= 0) return null;
  return normalizePortfolioTransaction({
    id: asString(row.id || `legacy_${symbol}`),
    symbol,
    assetName: asString(row.assetName || symbol),
    description: row.description,
    notes: row.notes,
    purchaseType: row.purchaseType || 'Delivery',
    tradeDate: row.purchaseDate,
    platform: row.platform,
    country: row.country || 'India',
    state: row.state,
    segment: row.segment || 'Equity',
    quantity,
    price,
    currentPrice: asNumber(row.currentPrice, price),
    currencySymbol: row.currencySymbol || 'INR',
    chargeSnapshot: {
      platformId: '',
      platformLabel: asString(row.platform),
      segmentKey: '',
      segmentLabel: '',
      side: 'BUY',
      turnover: round(quantity * price),
      lines: [],
      totals: {},
      totalCharges: 0,
      sourceTitle: 'Legacy migrated holding',
      sourceUrl: '',
      exactness: 'legacy_import',
      registryVersion: '',
    },
    legacyImported: true,
  });
};

export const normalizePortfolioMap = (input) => {
  const next = {};
  if (input && typeof input === 'object' && !Array.isArray(input)) {
    Object.entries(input).forEach(([name, rows]) => {
      const portfolioName = asString(name);
      if (!portfolioName) return;
      next[portfolioName] = (Array.isArray(rows) ? rows : [])
        .map((row) => {
          const isTransaction = row && typeof row === 'object' && (row.entryType || row.side || row.transactionType);
          return isTransaction ? normalizePortfolioTransaction(row) : migrateLegacyPosition(row);
        })
        .filter(Boolean);
    });
  }
  return Object.keys(next).length ? next : { Main: [] };
};

export const normalizeLegacyPortfolioRows = (rows) => ({
  Main: Array.isArray(rows) ? rows.map((row) => migrateLegacyPosition(row)).filter(Boolean) : [],
});

export const deriveHoldingsFromTransactions = (transactions) => {
  const rows = Array.isArray(transactions) ? transactions.map((row) => normalizePortfolioTransaction(row)).filter(Boolean) : [];
  const sorted = [...rows].sort((a, b) =>
    `${a.tradeDate || a.createdAt}|${a.createdAt}|${a.id}`.localeCompare(`${b.tradeDate || b.createdAt}|${b.createdAt}|${b.id}`));
  const grouped = new Map();

  sorted.forEach((txn) => {
    const key = txn.symbol;
    const current = grouped.get(key) || {
      id: key,
      symbol: key,
      assetName: txn.assetName || key,
      description: txn.description || '',
      notes: txn.notes || '',
      purchaseType: txn.purchaseType || 'Delivery',
      segment: txn.segment || 'Equity',
      platform: txn.platform || '',
      country: txn.country || 'India',
      currencySymbol: txn.currencySymbol || 'INR',
      currentPrice: txn.currentPrice || txn.price || 0,
      quantity: 0,
      costBasis: 0,
      realizedPnl: 0,
      totalChargesPaid: 0,
      transactionCount: 0,
      lastTradeDate: '',
      transactions: [],
    };
    const charges = asNumber(txn.chargeSnapshot?.totalCharges, 0);
    current.assetName = txn.assetName || current.assetName;
    current.description = txn.description || current.description;
    current.notes = txn.notes || current.notes;
    current.purchaseType = txn.purchaseType || current.purchaseType;
    current.segment = txn.segment || current.segment;
    current.platform = txn.platform || current.platform;
    current.country = txn.country || current.country;
    current.currencySymbol = txn.currencySymbol || current.currencySymbol;
    current.currentPrice = asNumber(txn.currentPrice, current.currentPrice);
    current.lastTradeDate = txn.tradeDate || txn.createdAt || current.lastTradeDate;
    current.totalChargesPaid += charges;
    current.transactionCount += 1;
    current.transactions.push(txn);

    if (txn.side === 'BUY') {
      current.quantity += txn.quantity;
      current.costBasis += (txn.quantity * txn.price) + charges;
    } else if (txn.side === 'SELL') {
      const sellQty = Math.min(txn.quantity, current.quantity);
      const avgCost = current.quantity > 0 ? current.costBasis / current.quantity : 0;
      const costRemoved = avgCost * sellQty;
      const netSaleProceeds = (sellQty * txn.price) - charges;
      current.realizedPnl += netSaleProceeds - costRemoved;
      current.quantity = Math.max(current.quantity - sellQty, 0);
      current.costBasis = Math.max(current.costBasis - costRemoved, 0);
    } else if (txn.side === 'DIVIDEND') {
      current.realizedPnl += (txn.quantity * txn.price) - charges;
    } else if (txn.side === 'FEE' || txn.side === 'TAX') {
      current.realizedPnl -= charges || (txn.quantity * txn.price);
    } else if (txn.side === 'ADJUSTMENT') {
      const subtype = asString(txn.transactionSubtype).toLowerCase();
      if ((subtype === 'split' || subtype === 'bonus') && txn.quantity > 0) {
        current.quantity += txn.quantity;
      } else {
        current.costBasis += txn.quantity * txn.price;
      }
    }

    grouped.set(key, current);
  });

  return [...grouped.values()].map((row) => {
    const averageCost = row.quantity > 0 ? row.costBasis / row.quantity : 0;
    const current = row.currentPrice * row.quantity;
    const projectedExitCharges = row.quantity > 0 ? round(current * 0.0015) : 0;
    const grossPnl = current - row.costBasis;
    const netPnl = grossPnl - projectedExitCharges;
    return {
      id: row.id,
      symbol: row.symbol,
      assetName: row.assetName,
      description: row.description,
      notes: row.notes,
      purchaseType: row.purchaseType,
      segment: row.segment,
      platform: row.platform,
      country: row.country,
      currencySymbol: row.currencySymbol,
      quantity: round(row.quantity, 6),
      buyPrice: round(averageCost, 4),
      averageCost: round(averageCost, 4),
      invested: round(row.costBasis),
      currentPrice: round(row.currentPrice, 4),
      current: round(current),
      grossPnl: round(grossPnl),
      pnl: round(netPnl),
      netPnl: round(netPnl),
      projectedExitCharges: round(projectedExitCharges),
      realizedPnl: round(row.realizedPnl),
      totalChargesPaid: round(row.totalChargesPaid),
      transactionCount: row.transactionCount,
      lastTradeDate: row.lastTradeDate,
      transactions: row.transactions,
    };
  }).sort((a, b) => b.current - a.current);
};

export const getPortfolioStats = (holdingsInput, transactionsInput = []) => {
  const rows = Array.isArray(holdingsInput) ? holdingsInput : deriveHoldingsFromTransactions(transactionsInput);
  const bySegment = {};
  const byPlatform = {};
  const byCountry = {};
  const totals = rows.reduce((acc, row) => {
    const segment = asString(row.segment || 'Other') || 'Other';
    const platform = asString(row.platform || 'Unspecified') || 'Unspecified';
    const country = asString(row.country || 'Unspecified') || 'Unspecified';
    if (!bySegment[segment]) bySegment[segment] = { invested: 0, current: 0, net: 0 };
    if (!byPlatform[platform]) byPlatform[platform] = { invested: 0, current: 0, net: 0 };
    if (!byCountry[country]) byCountry[country] = { invested: 0, current: 0, net: 0 };
    bySegment[segment].invested += asNumber(row.invested, 0);
    bySegment[segment].current += asNumber(row.current, 0);
    bySegment[segment].net += asNumber(row.netPnl ?? row.pnl, 0);
    byPlatform[platform].invested += asNumber(row.invested, 0);
    byPlatform[platform].current += asNumber(row.current, 0);
    byPlatform[platform].net += asNumber(row.netPnl ?? row.pnl, 0);
    byCountry[country].invested += asNumber(row.invested, 0);
    byCountry[country].current += asNumber(row.current, 0);
    byCountry[country].net += asNumber(row.netPnl ?? row.pnl, 0);
    acc.invested += asNumber(row.invested, 0);
    acc.current += asNumber(row.current, 0);
    acc.realizedPnl += asNumber(row.realizedPnl, 0);
    acc.platformFees += asNumber(row.totalChargesPaid, 0);
    acc.tax += asNumber(row.projectedExitCharges, 0);
    acc.profitable += asNumber(row.netPnl ?? row.pnl, 0) >= 0 ? 1 : 0;
    return acc;
  }, { invested: 0, current: 0, tax: 0, platformFees: 0, profitable: 0, realizedPnl: 0 });

  const mapToRows = (grouped) => Object.keys(grouped).map((label) => ({
    label,
    invested: round(grouped[label].invested),
    current: round(grouped[label].current),
    pnl: round(grouped[label].current - grouped[label].invested),
    net: round(grouped[label].net),
  }));

  const recentPurchases = (Array.isArray(transactionsInput) ? transactionsInput : [])
    .map((row) => normalizePortfolioTransaction(row))
    .filter(Boolean)
    .sort((a, b) => `${b.tradeDate || b.createdAt}`.localeCompare(`${a.tradeDate || a.createdAt}`))
    .slice(0, 6);

  const grossPnl = totals.current - totals.invested;
  return {
    holdings: rows.length,
    profitable: totals.profitable,
    invested: round(totals.invested),
    current: round(totals.current),
    tax: round(totals.tax),
    platformFees: round(totals.platformFees),
    realizedPnl: round(totals.realizedPnl),
    grossPnl: round(grossPnl),
    netAfterCosts: round(grossPnl - totals.tax),
    bySegment: mapToRows(bySegment),
    byPlatform: mapToRows(byPlatform),
    byCountry: mapToRows(byCountry),
    recentPurchases,
    holdingsRows: rows,
  };
};

export const normalizePortfolioSnapshots = (raw) => (Array.isArray(raw) ? raw : [])
  .map((row) => {
    if (!row || typeof row !== 'object') return null;
    const dateKey = asString(row.dateKey);
    const capturedAt = asString(row.capturedAt);
    const overall = row.overall && typeof row.overall === 'object' ? row.overall : {};
    const portfolios = Array.isArray(row.portfolios) ? row.portfolios : [];
    return {
      dateKey: dateKey || capturedAt.slice(0, 10),
      capturedAt: capturedAt || `${dateKey}T00:00:00.000Z`,
      overall: {
        holdings: asNumber(overall.holdings, 0),
        invested: asNumber(overall.invested, 0),
        current: asNumber(overall.current, 0),
        grossPnl: asNumber(overall.grossPnl, 0),
        netAfterCosts: asNumber(overall.netAfterCosts, 0),
      },
      portfolios: portfolios
        .map((item) => ({
          name: asString(item?.name),
          holdings: asNumber(item?.holdings, 0),
          invested: asNumber(item?.invested, 0),
          current: asNumber(item?.current, 0),
          pnl: asNumber(item?.pnl, 0),
          net: asNumber(item?.net, 0),
        }))
        .filter((item) => item.name),
    };
  })
  .filter(Boolean)
  .sort((a, b) => `${a.capturedAt || a.dateKey}`.localeCompare(`${b.capturedAt || b.dateKey}`));

export const buildPortfolioSnapshot = (portfolioMap, capturedAt = new Date().toISOString()) => {
  const normalized = normalizePortfolioMap(portfolioMap || {});
  const portfolioRows = Object.keys(normalized).map((name) => {
    const holdings = deriveHoldingsFromTransactions(normalized[name] || []);
    const stats = getPortfolioStats(holdings, normalized[name] || []);
    return {
      name,
      holdings: stats.holdings,
      invested: round(stats.invested),
      current: round(stats.current),
      pnl: round(stats.grossPnl),
      net: round(stats.netAfterCosts),
    };
  });
  const allTransactions = Object.values(normalized).flat();
  const allHoldings = deriveHoldingsFromTransactions(allTransactions);
  const overall = getPortfolioStats(allHoldings, allTransactions);
  return {
    dateKey: capturedAt.slice(0, 10),
    capturedAt,
    overall: {
      holdings: overall.holdings,
      invested: round(overall.invested),
      current: round(overall.current),
      grossPnl: round(overall.grossPnl),
      netAfterCosts: round(overall.netAfterCosts),
    },
    portfolios: portfolioRows,
  };
};
